
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../metadata/">
      
      
        <link rel="next" href="../create_template/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.49">
    
    
      
        <title>StructuralVariant - pyphetools</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.6f8fc17f.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
      
        <meta  property="og:type"  content="website" >
      
        <meta  property="og:title"  content="StructuralVariant - pyphetools" >
      
        <meta  property="og:description"  content="None" >
      
        <meta  property="og:image"  content="./assets/images/social/api/creation/structural_variant.png" >
      
        <meta  property="og:image:type"  content="image/png" >
      
        <meta  property="og:image:width"  content="1200" >
      
        <meta  property="og:image:height"  content="630" >
      
        <meta  property="og:url"  content="None" >
      
        <meta  name="twitter:card"  content="summary_large_image" >
      
        <meta  name="twitter:title"  content="StructuralVariant - pyphetools" >
      
        <meta  name="twitter:description"  content="None" >
      
        <meta  name="twitter:image"  content="./assets/images/social/api/creation/structural_variant.png" >
      
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="white">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#structuralvariant" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="pyphetools" class="md-header__button md-logo" aria-label="pyphetools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            pyphetools
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              StructuralVariant
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="white"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../user-guide/template/" class="md-tabs__link">
          
  
  User guide

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../visualization/" class="md-tabs__link">
          
  
  Visualization

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../developers/developers/" class="md-tabs__link">
          
  
  Developers

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../overview/" class="md-tabs__link">
          
  
  API

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="pyphetools" class="md-nav__button md-logo" aria-label="pyphetools" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    pyphetools
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    User guide
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            User guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user-guide/template/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data-Entry Template
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user-guide/excel/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Excel Template
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user-guide/python_notebook/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Python (Jupyter) notebook for the Excel template
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user-guide/summary/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user-guide/tips_for_curation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tips for accurate curation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user-guide/variant_notation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant Notation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user-guide/discombobulator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Discombobulator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user-guide/variant_column_mapper.md" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant column mapper
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../visualization/kaplan_meier_visualizer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Kaplan Meier
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Developers
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Developers
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../developers/developers/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    For developers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../developers/installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../developers/hpoa_editing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    HPOA files
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../developers/internal/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Internal
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    API
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" checked>
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    creation
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            creation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../case_template_encoder/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CaseTemplateEncoder
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../citation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Citation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cohort_encoder/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CohortEncoder
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../discombobulator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Discombobulator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../disease/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Disease
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../hgvs_variant/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    HgvsVariant
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../hpo_cr/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    HpoConceptRecognizer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../hpo_exact_cr/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    HpoExactConceptRecognizer
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../hpo_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    HpoParser
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../hp_term/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    HpTerm
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../individual/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Individual
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MetaData
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    StructuralVariant
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    StructuralVariant
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#pyphetools.creation.StructuralVariant" class="md-nav__link">
    <span class="md-ellipsis">
      StructuralVariant
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyphetools.creation.StructuralVariant.chromosomal_deletion" class="md-nav__link">
    <span class="md-ellipsis">
      chromosomal_deletion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyphetools.creation.StructuralVariant.chromosomal_duplication" class="md-nav__link">
    <span class="md-ellipsis">
      chromosomal_duplication
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyphetools.creation.StructuralVariant.chromosomal_inversion" class="md-nav__link">
    <span class="md-ellipsis">
      chromosomal_inversion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyphetools.creation.StructuralVariant.chromosomal_translocation" class="md-nav__link">
    <span class="md-ellipsis">
      chromosomal_translocation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyphetools.creation.StructuralVariant.to_ga4gh_variant_interpretation" class="md-nav__link">
    <span class="md-ellipsis">
      to_ga4gh_variant_interpretation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pyphetools.creation.StructuralVariant.to_variant_interpretation_202" class="md-nav__link">
    <span class="md-ellipsis">
      to_variant_interpretation_202
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../create_template/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    TemplateCreator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../thresholder/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Thresholder
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../thresholded_column_mapper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ThresholdedColumnMapper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_column_mapper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    VariantColumnMapper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_manager/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    VariantManager
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_validator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    VariantValidator
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    validation
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            validation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/cohort_validator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CohortValidator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/content_validator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ContentValidator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/ontology_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    OntologyQC
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/validation_result/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ValidationResult
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" >
        
          
          <label class="md-nav__link" for="__nav_5_4" id="__nav_5_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    visualization
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/detailed_suppl_table/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    DetailedSupplTable
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/hpoa_table_creator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    HpoaTableCreator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/onset_calculator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    OnsetCalculator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/phenopacket_ingestor/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PhenopacketIngestor
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/phenopacket_table/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PhenopacketTable
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../visualization/qc_visualizer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QcVisualizer
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="structuralvariant">StructuralVariant</h1>


<div class="doc doc-object doc-class">



<a id="pyphetools.creation.StructuralVariant"></a>
    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="pyphetools.creation.variant.Variant" href="../variant/#pyphetools.creation.Variant">Variant</a></code></p>


        <p>This encapsulates variant about a structural variant
For instance, we may see things like this
chr7.hg19:g.(35674000_37280000)_(46_111_000_46_598_000)del (fuzzy boundaries)
chr7.hg19:g(38521704_45810267)del (precise boundaries)
rsa7p14.1(kit P179)x1
46,XY.ish del(7)(p14.1)(RP11-816F16-)
46,XX.ish del(7)(p14.1p14.1)(GLI3-)
46,XY.ish del(7)(p14.1)(GLI3-)[56]/7p14.1(GLI3x2)[44] (various ICSN or bespoke)
Our strategy is not to model these variants precisely. Instead, for genotype-phenotype
analysis, it may be enough to know that these are structural variants and thus
likely to be complete loss of function.
So we record something about the variant and add it by hand to the Individual object
We want to be able to create a GA4GH VariationDescriptor object with the following fields
- id - required, autogenerate if user provides no id
- variant (VRS) -- leave this field empty
- label -- the original contents of the cell, e.g., 46,XY.ish del(7)(p14.1)(RP11-816F16-)
- gene context -- the gene that is disrupted by the structural variant
- expressions, empty
- vcf_record, empty
- structural type: Ontology Class
- allelic state: het/hom/emi etc.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>cell_contents</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the string from the original table that we want to map as a structural variant</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_symbol</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the gene affected by the structural variant, e.g., GLI3</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sequence_ontology_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An identifier from the Sequence Ontology</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sequence_ontology_label</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the SO label corresponding to the ID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_id</code>
            </td>
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>an identifier for the variant, optional</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>






              <details class="quote">
                <summary>Source code in <code>pyphetools/creation/structural_variant.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-9">  9</a></span>
<span class="normal"><a href="#__codelineno-0-10"> 10</a></span>
<span class="normal"><a href="#__codelineno-0-11"> 11</a></span>
<span class="normal"><a href="#__codelineno-0-12"> 12</a></span>
<span class="normal"><a href="#__codelineno-0-13"> 13</a></span>
<span class="normal"><a href="#__codelineno-0-14"> 14</a></span>
<span class="normal"><a href="#__codelineno-0-15"> 15</a></span>
<span class="normal"><a href="#__codelineno-0-16"> 16</a></span>
<span class="normal"><a href="#__codelineno-0-17"> 17</a></span>
<span class="normal"><a href="#__codelineno-0-18"> 18</a></span>
<span class="normal"><a href="#__codelineno-0-19"> 19</a></span>
<span class="normal"><a href="#__codelineno-0-20"> 20</a></span>
<span class="normal"><a href="#__codelineno-0-21"> 21</a></span>
<span class="normal"><a href="#__codelineno-0-22"> 22</a></span>
<span class="normal"><a href="#__codelineno-0-23"> 23</a></span>
<span class="normal"><a href="#__codelineno-0-24"> 24</a></span>
<span class="normal"><a href="#__codelineno-0-25"> 25</a></span>
<span class="normal"><a href="#__codelineno-0-26"> 26</a></span>
<span class="normal"><a href="#__codelineno-0-27"> 27</a></span>
<span class="normal"><a href="#__codelineno-0-28"> 28</a></span>
<span class="normal"><a href="#__codelineno-0-29"> 29</a></span>
<span class="normal"><a href="#__codelineno-0-30"> 30</a></span>
<span class="normal"><a href="#__codelineno-0-31"> 31</a></span>
<span class="normal"><a href="#__codelineno-0-32"> 32</a></span>
<span class="normal"><a href="#__codelineno-0-33"> 33</a></span>
<span class="normal"><a href="#__codelineno-0-34"> 34</a></span>
<span class="normal"><a href="#__codelineno-0-35"> 35</a></span>
<span class="normal"><a href="#__codelineno-0-36"> 36</a></span>
<span class="normal"><a href="#__codelineno-0-37"> 37</a></span>
<span class="normal"><a href="#__codelineno-0-38"> 38</a></span>
<span class="normal"><a href="#__codelineno-0-39"> 39</a></span>
<span class="normal"><a href="#__codelineno-0-40"> 40</a></span>
<span class="normal"><a href="#__codelineno-0-41"> 41</a></span>
<span class="normal"><a href="#__codelineno-0-42"> 42</a></span>
<span class="normal"><a href="#__codelineno-0-43"> 43</a></span>
<span class="normal"><a href="#__codelineno-0-44"> 44</a></span>
<span class="normal"><a href="#__codelineno-0-45"> 45</a></span>
<span class="normal"><a href="#__codelineno-0-46"> 46</a></span>
<span class="normal"><a href="#__codelineno-0-47"> 47</a></span>
<span class="normal"><a href="#__codelineno-0-48"> 48</a></span>
<span class="normal"><a href="#__codelineno-0-49"> 49</a></span>
<span class="normal"><a href="#__codelineno-0-50"> 50</a></span>
<span class="normal"><a href="#__codelineno-0-51"> 51</a></span>
<span class="normal"><a href="#__codelineno-0-52"> 52</a></span>
<span class="normal"><a href="#__codelineno-0-53"> 53</a></span>
<span class="normal"><a href="#__codelineno-0-54"> 54</a></span>
<span class="normal"><a href="#__codelineno-0-55"> 55</a></span>
<span class="normal"><a href="#__codelineno-0-56"> 56</a></span>
<span class="normal"><a href="#__codelineno-0-57"> 57</a></span>
<span class="normal"><a href="#__codelineno-0-58"> 58</a></span>
<span class="normal"><a href="#__codelineno-0-59"> 59</a></span>
<span class="normal"><a href="#__codelineno-0-60"> 60</a></span>
<span class="normal"><a href="#__codelineno-0-61"> 61</a></span>
<span class="normal"><a href="#__codelineno-0-62"> 62</a></span>
<span class="normal"><a href="#__codelineno-0-63"> 63</a></span>
<span class="normal"><a href="#__codelineno-0-64"> 64</a></span>
<span class="normal"><a href="#__codelineno-0-65"> 65</a></span>
<span class="normal"><a href="#__codelineno-0-66"> 66</a></span>
<span class="normal"><a href="#__codelineno-0-67"> 67</a></span>
<span class="normal"><a href="#__codelineno-0-68"> 68</a></span>
<span class="normal"><a href="#__codelineno-0-69"> 69</a></span>
<span class="normal"><a href="#__codelineno-0-70"> 70</a></span>
<span class="normal"><a href="#__codelineno-0-71"> 71</a></span>
<span class="normal"><a href="#__codelineno-0-72"> 72</a></span>
<span class="normal"><a href="#__codelineno-0-73"> 73</a></span>
<span class="normal"><a href="#__codelineno-0-74"> 74</a></span>
<span class="normal"><a href="#__codelineno-0-75"> 75</a></span>
<span class="normal"><a href="#__codelineno-0-76"> 76</a></span>
<span class="normal"><a href="#__codelineno-0-77"> 77</a></span>
<span class="normal"><a href="#__codelineno-0-78"> 78</a></span>
<span class="normal"><a href="#__codelineno-0-79"> 79</a></span>
<span class="normal"><a href="#__codelineno-0-80"> 80</a></span>
<span class="normal"><a href="#__codelineno-0-81"> 81</a></span>
<span class="normal"><a href="#__codelineno-0-82"> 82</a></span>
<span class="normal"><a href="#__codelineno-0-83"> 83</a></span>
<span class="normal"><a href="#__codelineno-0-84"> 84</a></span>
<span class="normal"><a href="#__codelineno-0-85"> 85</a></span>
<span class="normal"><a href="#__codelineno-0-86"> 86</a></span>
<span class="normal"><a href="#__codelineno-0-87"> 87</a></span>
<span class="normal"><a href="#__codelineno-0-88"> 88</a></span>
<span class="normal"><a href="#__codelineno-0-89"> 89</a></span>
<span class="normal"><a href="#__codelineno-0-90"> 90</a></span>
<span class="normal"><a href="#__codelineno-0-91"> 91</a></span>
<span class="normal"><a href="#__codelineno-0-92"> 92</a></span>
<span class="normal"><a href="#__codelineno-0-93"> 93</a></span>
<span class="normal"><a href="#__codelineno-0-94"> 94</a></span>
<span class="normal"><a href="#__codelineno-0-95"> 95</a></span>
<span class="normal"><a href="#__codelineno-0-96"> 96</a></span>
<span class="normal"><a href="#__codelineno-0-97"> 97</a></span>
<span class="normal"><a href="#__codelineno-0-98"> 98</a></span>
<span class="normal"><a href="#__codelineno-0-99"> 99</a></span>
<span class="normal"><a href="#__codelineno-0-100">100</a></span>
<span class="normal"><a href="#__codelineno-0-101">101</a></span>
<span class="normal"><a href="#__codelineno-0-102">102</a></span>
<span class="normal"><a href="#__codelineno-0-103">103</a></span>
<span class="normal"><a href="#__codelineno-0-104">104</a></span>
<span class="normal"><a href="#__codelineno-0-105">105</a></span>
<span class="normal"><a href="#__codelineno-0-106">106</a></span>
<span class="normal"><a href="#__codelineno-0-107">107</a></span>
<span class="normal"><a href="#__codelineno-0-108">108</a></span>
<span class="normal"><a href="#__codelineno-0-109">109</a></span>
<span class="normal"><a href="#__codelineno-0-110">110</a></span>
<span class="normal"><a href="#__codelineno-0-111">111</a></span>
<span class="normal"><a href="#__codelineno-0-112">112</a></span>
<span class="normal"><a href="#__codelineno-0-113">113</a></span>
<span class="normal"><a href="#__codelineno-0-114">114</a></span>
<span class="normal"><a href="#__codelineno-0-115">115</a></span>
<span class="normal"><a href="#__codelineno-0-116">116</a></span>
<span class="normal"><a href="#__codelineno-0-117">117</a></span>
<span class="normal"><a href="#__codelineno-0-118">118</a></span>
<span class="normal"><a href="#__codelineno-0-119">119</a></span>
<span class="normal"><a href="#__codelineno-0-120">120</a></span>
<span class="normal"><a href="#__codelineno-0-121">121</a></span>
<span class="normal"><a href="#__codelineno-0-122">122</a></span>
<span class="normal"><a href="#__codelineno-0-123">123</a></span>
<span class="normal"><a href="#__codelineno-0-124">124</a></span>
<span class="normal"><a href="#__codelineno-0-125">125</a></span>
<span class="normal"><a href="#__codelineno-0-126">126</a></span>
<span class="normal"><a href="#__codelineno-0-127">127</a></span>
<span class="normal"><a href="#__codelineno-0-128">128</a></span>
<span class="normal"><a href="#__codelineno-0-129">129</a></span>
<span class="normal"><a href="#__codelineno-0-130">130</a></span>
<span class="normal"><a href="#__codelineno-0-131">131</a></span>
<span class="normal"><a href="#__codelineno-0-132">132</a></span>
<span class="normal"><a href="#__codelineno-0-133">133</a></span>
<span class="normal"><a href="#__codelineno-0-134">134</a></span>
<span class="normal"><a href="#__codelineno-0-135">135</a></span>
<span class="normal"><a href="#__codelineno-0-136">136</a></span>
<span class="normal"><a href="#__codelineno-0-137">137</a></span>
<span class="normal"><a href="#__codelineno-0-138">138</a></span>
<span class="normal"><a href="#__codelineno-0-139">139</a></span>
<span class="normal"><a href="#__codelineno-0-140">140</a></span>
<span class="normal"><a href="#__codelineno-0-141">141</a></span>
<span class="normal"><a href="#__codelineno-0-142">142</a></span>
<span class="normal"><a href="#__codelineno-0-143">143</a></span>
<span class="normal"><a href="#__codelineno-0-144">144</a></span>
<span class="normal"><a href="#__codelineno-0-145">145</a></span>
<span class="normal"><a href="#__codelineno-0-146">146</a></span>
<span class="normal"><a href="#__codelineno-0-147">147</a></span>
<span class="normal"><a href="#__codelineno-0-148">148</a></span>
<span class="normal"><a href="#__codelineno-0-149">149</a></span>
<span class="normal"><a href="#__codelineno-0-150">150</a></span>
<span class="normal"><a href="#__codelineno-0-151">151</a></span>
<span class="normal"><a href="#__codelineno-0-152">152</a></span>
<span class="normal"><a href="#__codelineno-0-153">153</a></span>
<span class="normal"><a href="#__codelineno-0-154">154</a></span>
<span class="normal"><a href="#__codelineno-0-155">155</a></span>
<span class="normal"><a href="#__codelineno-0-156">156</a></span>
<span class="normal"><a href="#__codelineno-0-157">157</a></span>
<span class="normal"><a href="#__codelineno-0-158">158</a></span>
<span class="normal"><a href="#__codelineno-0-159">159</a></span>
<span class="normal"><a href="#__codelineno-0-160">160</a></span>
<span class="normal"><a href="#__codelineno-0-161">161</a></span>
<span class="normal"><a href="#__codelineno-0-162">162</a></span>
<span class="normal"><a href="#__codelineno-0-163">163</a></span>
<span class="normal"><a href="#__codelineno-0-164">164</a></span>
<span class="normal"><a href="#__codelineno-0-165">165</a></span>
<span class="normal"><a href="#__codelineno-0-166">166</a></span>
<span class="normal"><a href="#__codelineno-0-167">167</a></span>
<span class="normal"><a href="#__codelineno-0-168">168</a></span>
<span class="normal"><a href="#__codelineno-0-169">169</a></span>
<span class="normal"><a href="#__codelineno-0-170">170</a></span>
<span class="normal"><a href="#__codelineno-0-171">171</a></span>
<span class="normal"><a href="#__codelineno-0-172">172</a></span>
<span class="normal"><a href="#__codelineno-0-173">173</a></span>
<span class="normal"><a href="#__codelineno-0-174">174</a></span>
<span class="normal"><a href="#__codelineno-0-175">175</a></span>
<span class="normal"><a href="#__codelineno-0-176">176</a></span>
<span class="normal"><a href="#__codelineno-0-177">177</a></span>
<span class="normal"><a href="#__codelineno-0-178">178</a></span>
<span class="normal"><a href="#__codelineno-0-179">179</a></span>
<span class="normal"><a href="#__codelineno-0-180">180</a></span>
<span class="normal"><a href="#__codelineno-0-181">181</a></span>
<span class="normal"><a href="#__codelineno-0-182">182</a></span>
<span class="normal"><a href="#__codelineno-0-183">183</a></span>
<span class="normal"><a href="#__codelineno-0-184">184</a></span>
<span class="normal"><a href="#__codelineno-0-185">185</a></span>
<span class="normal"><a href="#__codelineno-0-186">186</a></span>
<span class="normal"><a href="#__codelineno-0-187">187</a></span>
<span class="normal"><a href="#__codelineno-0-188">188</a></span>
<span class="normal"><a href="#__codelineno-0-189">189</a></span>
<span class="normal"><a href="#__codelineno-0-190">190</a></span>
<span class="normal"><a href="#__codelineno-0-191">191</a></span>
<span class="normal"><a href="#__codelineno-0-192">192</a></span>
<span class="normal"><a href="#__codelineno-0-193">193</a></span>
<span class="normal"><a href="#__codelineno-0-194">194</a></span>
<span class="normal"><a href="#__codelineno-0-195">195</a></span>
<span class="normal"><a href="#__codelineno-0-196">196</a></span>
<span class="normal"><a href="#__codelineno-0-197">197</a></span>
<span class="normal"><a href="#__codelineno-0-198">198</a></span>
<span class="normal"><a href="#__codelineno-0-199">199</a></span>
<span class="normal"><a href="#__codelineno-0-200">200</a></span>
<span class="normal"><a href="#__codelineno-0-201">201</a></span>
<span class="normal"><a href="#__codelineno-0-202">202</a></span>
<span class="normal"><a href="#__codelineno-0-203">203</a></span>
<span class="normal"><a href="#__codelineno-0-204">204</a></span>
<span class="normal"><a href="#__codelineno-0-205">205</a></span>
<span class="normal"><a href="#__codelineno-0-206">206</a></span>
<span class="normal"><a href="#__codelineno-0-207">207</a></span>
<span class="normal"><a href="#__codelineno-0-208">208</a></span>
<span class="normal"><a href="#__codelineno-0-209">209</a></span>
<span class="normal"><a href="#__codelineno-0-210">210</a></span>
<span class="normal"><a href="#__codelineno-0-211">211</a></span>
<span class="normal"><a href="#__codelineno-0-212">212</a></span>
<span class="normal"><a href="#__codelineno-0-213">213</a></span>
<span class="normal"><a href="#__codelineno-0-214">214</a></span>
<span class="normal"><a href="#__codelineno-0-215">215</a></span>
<span class="normal"><a href="#__codelineno-0-216">216</a></span>
<span class="normal"><a href="#__codelineno-0-217">217</a></span>
<span class="normal"><a href="#__codelineno-0-218">218</a></span>
<span class="normal"><a href="#__codelineno-0-219">219</a></span>
<span class="normal"><a href="#__codelineno-0-220">220</a></span>
<span class="normal"><a href="#__codelineno-0-221">221</a></span>
<span class="normal"><a href="#__codelineno-0-222">222</a></span>
<span class="normal"><a href="#__codelineno-0-223">223</a></span>
<span class="normal"><a href="#__codelineno-0-224">224</a></span>
<span class="normal"><a href="#__codelineno-0-225">225</a></span>
<span class="normal"><a href="#__codelineno-0-226">226</a></span>
<span class="normal"><a href="#__codelineno-0-227">227</a></span>
<span class="normal"><a href="#__codelineno-0-228">228</a></span>
<span class="normal"><a href="#__codelineno-0-229">229</a></span>
<span class="normal"><a href="#__codelineno-0-230">230</a></span>
<span class="normal"><a href="#__codelineno-0-231">231</a></span>
<span class="normal"><a href="#__codelineno-0-232">232</a></span>
<span class="normal"><a href="#__codelineno-0-233">233</a></span>
<span class="normal"><a href="#__codelineno-0-234">234</a></span>
<span class="normal"><a href="#__codelineno-0-235">235</a></span>
<span class="normal"><a href="#__codelineno-0-236">236</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-9" name="__codelineno-0-9"></a><span class="k">class</span><span class="w"> </span><span class="nc">StructuralVariant</span><span class="p">(</span><span class="n">Variant</span><span class="p">):</span>
<a id="__codelineno-0-10" name="__codelineno-0-10"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-11" name="__codelineno-0-11"></a><span class="sd">    This encapsulates variant about a structural variant</span>
<a id="__codelineno-0-12" name="__codelineno-0-12"></a><span class="sd">    For instance, we may see things like this</span>
<a id="__codelineno-0-13" name="__codelineno-0-13"></a><span class="sd">    chr7.hg19:g.(35674000_37280000)_(46_111_000_46_598_000)del (fuzzy boundaries)</span>
<a id="__codelineno-0-14" name="__codelineno-0-14"></a><span class="sd">    chr7.hg19:g(38521704_45810267)del (precise boundaries)</span>
<a id="__codelineno-0-15" name="__codelineno-0-15"></a><span class="sd">    rsa7p14.1(kit P179)x1</span>
<a id="__codelineno-0-16" name="__codelineno-0-16"></a><span class="sd">    46,XY.ish del(7)(p14.1)(RP11-816F16-)</span>
<a id="__codelineno-0-17" name="__codelineno-0-17"></a><span class="sd">    46,XX.ish del(7)(p14.1p14.1)(GLI3-)</span>
<a id="__codelineno-0-18" name="__codelineno-0-18"></a><span class="sd">    46,XY.ish del(7)(p14.1)(GLI3-)[56]/7p14.1(GLI3x2)[44] (various ICSN or bespoke)</span>
<a id="__codelineno-0-19" name="__codelineno-0-19"></a><span class="sd">    Our strategy is not to model these variants precisely. Instead, for genotype-phenotype</span>
<a id="__codelineno-0-20" name="__codelineno-0-20"></a><span class="sd">    analysis, it may be enough to know that these are structural variants and thus</span>
<a id="__codelineno-0-21" name="__codelineno-0-21"></a><span class="sd">    likely to be complete loss of function.</span>
<a id="__codelineno-0-22" name="__codelineno-0-22"></a><span class="sd">    So we record something about the variant and add it by hand to the Individual object</span>
<a id="__codelineno-0-23" name="__codelineno-0-23"></a><span class="sd">    We want to be able to create a GA4GH VariationDescriptor object with the following fields</span>
<a id="__codelineno-0-24" name="__codelineno-0-24"></a><span class="sd">    - id - required, autogenerate if user provides no id</span>
<a id="__codelineno-0-25" name="__codelineno-0-25"></a><span class="sd">    - variant (VRS) -- leave this field empty</span>
<a id="__codelineno-0-26" name="__codelineno-0-26"></a><span class="sd">    - label -- the original contents of the cell, e.g., 46,XY.ish del(7)(p14.1)(RP11-816F16-)</span>
<a id="__codelineno-0-27" name="__codelineno-0-27"></a><span class="sd">    - gene context -- the gene that is disrupted by the structural variant</span>
<a id="__codelineno-0-28" name="__codelineno-0-28"></a><span class="sd">    - expressions, empty</span>
<a id="__codelineno-0-29" name="__codelineno-0-29"></a><span class="sd">    - vcf_record, empty</span>
<a id="__codelineno-0-30" name="__codelineno-0-30"></a><span class="sd">    - structural type: Ontology Class</span>
<a id="__codelineno-0-31" name="__codelineno-0-31"></a><span class="sd">    - allelic state: het/hom/emi etc.</span>
<a id="__codelineno-0-32" name="__codelineno-0-32"></a>
<a id="__codelineno-0-33" name="__codelineno-0-33"></a><span class="sd">    :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-34" name="__codelineno-0-34"></a><span class="sd">    :type cell_contents: str</span>
<a id="__codelineno-0-35" name="__codelineno-0-35"></a><span class="sd">    :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-36" name="__codelineno-0-36"></a><span class="sd">    :type gene_symbol: str</span>
<a id="__codelineno-0-37" name="__codelineno-0-37"></a><span class="sd">    :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-38" name="__codelineno-0-38"></a><span class="sd">    :type gene_id: str</span>
<a id="__codelineno-0-39" name="__codelineno-0-39"></a><span class="sd">    :param sequence_ontology_id: An identifier from the Sequence Ontology</span>
<a id="__codelineno-0-40" name="__codelineno-0-40"></a><span class="sd">    :type sequence_ontology_id: str</span>
<a id="__codelineno-0-41" name="__codelineno-0-41"></a><span class="sd">    :param sequence_ontology_label: the SO label corresponding to the ID</span>
<a id="__codelineno-0-42" name="__codelineno-0-42"></a><span class="sd">    :type sequence_ontology_label: str</span>
<a id="__codelineno-0-43" name="__codelineno-0-43"></a><span class="sd">    :param variant_id: an identifier for the variant, optional</span>
<a id="__codelineno-0-44" name="__codelineno-0-44"></a><span class="sd">    :type variant_id: str</span>
<a id="__codelineno-0-45" name="__codelineno-0-45"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-46" name="__codelineno-0-46"></a>
<a id="__codelineno-0-47" name="__codelineno-0-47"></a>    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-48" name="__codelineno-0-48"></a>                 <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-49" name="__codelineno-0-49"></a>                 <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-50" name="__codelineno-0-50"></a>                 <span class="n">sequence_ontology_id</span><span class="p">,</span>
<a id="__codelineno-0-51" name="__codelineno-0-51"></a>                 <span class="n">sequence_ontology_label</span><span class="p">,</span>
<a id="__codelineno-0-52" name="__codelineno-0-52"></a>                 <span class="n">variant_id</span><span class="p">:</span> <span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-53" name="__codelineno-0-53"></a>        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
<a id="__codelineno-0-54" name="__codelineno-0-54"></a>        <span class="k">if</span> <span class="n">variant_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-55" name="__codelineno-0-55"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">_variant_id</span> <span class="o">=</span> <span class="s2">&quot;var_&quot;</span> <span class="o">+</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">string</span><span class="o">.</span><span class="n">ascii_letters</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">25</span><span class="p">))</span>
<a id="__codelineno-0-56" name="__codelineno-0-56"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-57" name="__codelineno-0-57"></a>            <span class="bp">self</span><span class="o">.</span><span class="n">_variant_id</span> <span class="o">=</span> <span class="n">variant_id</span>
<a id="__codelineno-0-58" name="__codelineno-0-58"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">_label</span> <span class="o">=</span> <span class="n">cell_contents</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
<a id="__codelineno-0-59" name="__codelineno-0-59"></a>        <span class="k">if</span> <span class="n">gene_symbol</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-60" name="__codelineno-0-60"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Need to pass a valid gene symbol!&quot;</span><span class="p">)</span>
<a id="__codelineno-0-61" name="__codelineno-0-61"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">_gene_symbol</span> <span class="o">=</span> <span class="n">gene_symbol</span>
<a id="__codelineno-0-62" name="__codelineno-0-62"></a>        <span class="k">if</span> <span class="n">gene_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-63" name="__codelineno-0-63"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Need to pass a valid HGNC gene id!&quot;</span><span class="p">)</span>
<a id="__codelineno-0-64" name="__codelineno-0-64"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">_hgnc_id</span> <span class="o">=</span> <span class="n">gene_id</span>
<a id="__codelineno-0-65" name="__codelineno-0-65"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">_so_id</span> <span class="o">=</span> <span class="n">sequence_ontology_id</span>
<a id="__codelineno-0-66" name="__codelineno-0-66"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">_so_label</span> <span class="o">=</span> <span class="n">sequence_ontology_label</span>
<a id="__codelineno-0-67" name="__codelineno-0-67"></a>        <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-68" name="__codelineno-0-68"></a>
<a id="__codelineno-0-69" name="__codelineno-0-69"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">to_ga4gh_variant_interpretation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">acmg</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-70" name="__codelineno-0-70"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-71" name="__codelineno-0-71"></a><span class="sd">        Transform this Variant object into a &quot;variantInterpretation&quot; message of the GA4GH Phenopacket schema</span>
<a id="__codelineno-0-72" name="__codelineno-0-72"></a><span class="sd">        &quot;&quot;&quot;</span>
<a id="__codelineno-0-73" name="__codelineno-0-73"></a>        <span class="n">vdescriptor</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">VariationDescriptor</span><span class="p">()</span>
<a id="__codelineno-0-74" name="__codelineno-0-74"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_variant_id</span>
<a id="__codelineno-0-75" name="__codelineno-0-75"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hgnc_id</span>
<a id="__codelineno-0-76" name="__codelineno-0-76"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_gene_symbol</span>
<a id="__codelineno-0-77" name="__codelineno-0-77"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_label</span>
<a id="__codelineno-0-78" name="__codelineno-0-78"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">molecule_context</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">MoleculeContext</span><span class="o">.</span><span class="n">genomic</span>
<a id="__codelineno-0-79" name="__codelineno-0-79"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-80" name="__codelineno-0-80"></a>            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="o">==</span> <span class="s1">&#39;heterozygous&#39;</span><span class="p">:</span>
<a id="__codelineno-0-81" name="__codelineno-0-81"></a>                <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;GENO:0000135&quot;</span>
<a id="__codelineno-0-82" name="__codelineno-0-82"></a>                <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;heterozygous&quot;</span>
<a id="__codelineno-0-83" name="__codelineno-0-83"></a>            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="o">==</span> <span class="s1">&#39;homozygous&#39;</span><span class="p">:</span>
<a id="__codelineno-0-84" name="__codelineno-0-84"></a>                <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;GENO:0000136&quot;</span>
<a id="__codelineno-0-85" name="__codelineno-0-85"></a>                <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;homozygous&quot;</span>
<a id="__codelineno-0-86" name="__codelineno-0-86"></a>            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="o">==</span> <span class="s1">&#39;hemizygous&#39;</span><span class="p">:</span>
<a id="__codelineno-0-87" name="__codelineno-0-87"></a>                <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;GENO:0000134&quot;</span>
<a id="__codelineno-0-88" name="__codelineno-0-88"></a>                <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;hemizygous&quot;</span>
<a id="__codelineno-0-89" name="__codelineno-0-89"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-90" name="__codelineno-0-90"></a>                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Did not recognize genotype </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-91" name="__codelineno-0-91"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">structural_type</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_so_id</span>
<a id="__codelineno-0-92" name="__codelineno-0-92"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">structural_type</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_so_label</span>
<a id="__codelineno-0-93" name="__codelineno-0-93"></a>        <span class="n">vinterpretation</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">VariantInterpretation</span><span class="p">()</span>
<a id="__codelineno-0-94" name="__codelineno-0-94"></a>        <span class="k">if</span> <span class="n">acmg</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-95" name="__codelineno-0-95"></a>            <span class="k">if</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;benign&#39;</span><span class="p">:</span>
<a id="__codelineno-0-96" name="__codelineno-0-96"></a>                <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">BENIGN</span>
<a id="__codelineno-0-97" name="__codelineno-0-97"></a>            <span class="k">elif</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span> <span class="o">==</span> <span class="s1">&#39;likely benign&#39;</span> <span class="ow">or</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;likely_benign&#39;</span><span class="p">:</span>
<a id="__codelineno-0-98" name="__codelineno-0-98"></a>                <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">LIKELY_BENIGN</span>
<a id="__codelineno-0-99" name="__codelineno-0-99"></a>            <span class="k">elif</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span> <span class="o">==</span> <span class="s1">&#39;uncertain significance&#39;</span> <span class="ow">or</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;uncertain_significance&#39;</span><span class="p">:</span>
<a id="__codelineno-0-100" name="__codelineno-0-100"></a>                <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">UNCERTAIN_SIGNIFICANCE</span>
<a id="__codelineno-0-101" name="__codelineno-0-101"></a>            <span class="k">elif</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span> <span class="o">==</span> <span class="s1">&#39;likely pathogenic&#39;</span> <span class="ow">or</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;likely_pathogenic&#39;</span><span class="p">:</span>
<a id="__codelineno-0-102" name="__codelineno-0-102"></a>                <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">LIKELY_PATHOGENIC</span>
<a id="__codelineno-0-103" name="__codelineno-0-103"></a>            <span class="k">elif</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span> <span class="o">==</span> <span class="s1">&#39;pathogenic&#39;</span> <span class="ow">or</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;pathogenic&#39;</span><span class="p">:</span>
<a id="__codelineno-0-104" name="__codelineno-0-104"></a>                <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">PATHOGENIC</span>
<a id="__codelineno-0-105" name="__codelineno-0-105"></a>            <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-106" name="__codelineno-0-106"></a>                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning- did not recognize ACMG category </span><span class="si">{</span><span class="n">acmg</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-107" name="__codelineno-0-107"></a>        <span class="n">vinterpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">CopyFrom</span><span class="p">(</span><span class="n">vdescriptor</span><span class="p">)</span>
<a id="__codelineno-0-108" name="__codelineno-0-108"></a>        <span class="k">return</span> <span class="n">vinterpretation</span>
<a id="__codelineno-0-109" name="__codelineno-0-109"></a>
<a id="__codelineno-0-110" name="__codelineno-0-110"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">to_variant_interpretation_202</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">acmg</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-111" name="__codelineno-0-111"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-112" name="__codelineno-0-112"></a><span class="sd">        Transform this Variant object into a &quot;variantInterpretation&quot; message of the GA4GH Phenopacket schema</span>
<a id="__codelineno-0-113" name="__codelineno-0-113"></a><span class="sd">        &quot;&quot;&quot;</span>
<a id="__codelineno-0-114" name="__codelineno-0-114"></a>        <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">VariantInterpretation</span> <span class="k">as</span> <span class="n">VariantInterpretation202</span>
<a id="__codelineno-0-115" name="__codelineno-0-115"></a>        <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">VariationDescriptor</span> <span class="k">as</span> <span class="n">VariationDescriptor202</span>
<a id="__codelineno-0-116" name="__codelineno-0-116"></a>        <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">GeneDescriptor</span> <span class="k">as</span> <span class="n">GeneDescriptor202</span>
<a id="__codelineno-0-117" name="__codelineno-0-117"></a>        <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">OntologyClass</span> <span class="k">as</span> <span class="n">OntologyClass202</span>
<a id="__codelineno-0-118" name="__codelineno-0-118"></a>        <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">MoleculeContext</span> <span class="k">as</span> <span class="n">MoleculeContext202</span>
<a id="__codelineno-0-119" name="__codelineno-0-119"></a>
<a id="__codelineno-0-120" name="__codelineno-0-120"></a>        <span class="n">gdesc</span> <span class="o">=</span> <span class="n">GeneDescriptor202</span><span class="p">(</span><span class="n">value_id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_hgnc_id</span><span class="p">,</span> <span class="n">symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_gene_symbol</span><span class="p">)</span>
<a id="__codelineno-0-121" name="__codelineno-0-121"></a>        <span class="n">vdescriptor</span> <span class="o">=</span> <span class="n">VariationDescriptor202</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_variant_id</span><span class="p">,</span>
<a id="__codelineno-0-122" name="__codelineno-0-122"></a>                                             <span class="n">gene_context</span><span class="o">=</span><span class="n">gdesc</span><span class="p">,</span>
<a id="__codelineno-0-123" name="__codelineno-0-123"></a>                                             <span class="n">label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_label</span><span class="p">,</span>
<a id="__codelineno-0-124" name="__codelineno-0-124"></a>                                             <span class="n">molecule_context</span><span class="o">=</span><span class="n">MoleculeContext202</span><span class="o">.</span><span class="n">genomic</span><span class="p">)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125"></a>        <span class="c1"># Note that it is possible to first create a StructuralVariant object and later set its genotype</span>
<a id="__codelineno-0-126" name="__codelineno-0-126"></a>        <span class="c1"># Therefore, it is not an error if gt_term is None</span>
<a id="__codelineno-0-127" name="__codelineno-0-127"></a>        <span class="n">gt_term</span> <span class="o">=</span> <span class="n">Variant</span><span class="o">.</span><span class="n">_get_genotype_term</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span><span class="p">)</span>
<a id="__codelineno-0-128" name="__codelineno-0-128"></a>        <span class="k">if</span> <span class="n">gt_term</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-129" name="__codelineno-0-129"></a>            <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span> <span class="o">=</span> <span class="n">gt_term</span>
<a id="__codelineno-0-130" name="__codelineno-0-130"></a>        <span class="n">structural_type</span> <span class="o">=</span> <span class="n">OntologyClass202</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_so_id</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_so_label</span><span class="p">)</span>
<a id="__codelineno-0-131" name="__codelineno-0-131"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">structural_type</span> <span class="o">=</span> <span class="n">structural_type</span>
<a id="__codelineno-0-132" name="__codelineno-0-132"></a>        <span class="n">vinterpretation</span> <span class="o">=</span> <span class="n">VariantInterpretation202</span><span class="p">(</span><span class="n">variation_descriptor</span><span class="o">=</span><span class="n">vdescriptor</span><span class="p">)</span>
<a id="__codelineno-0-133" name="__codelineno-0-133"></a>        <span class="n">acmg_code</span> <span class="o">=</span> <span class="n">Variant</span><span class="o">.</span><span class="n">_get_acmg_classification</span><span class="p">(</span><span class="n">acmg</span><span class="o">=</span><span class="n">acmg</span><span class="p">)</span>
<a id="__codelineno-0-134" name="__codelineno-0-134"></a>        <span class="k">if</span> <span class="n">acmg_code</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-135" name="__codelineno-0-135"></a>            <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">acmg_code</span>
<a id="__codelineno-0-136" name="__codelineno-0-136"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-137" name="__codelineno-0-137"></a>            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning- did not recognize ACMG category </span><span class="se">\&quot;</span><span class="si">{</span><span class="n">acmg</span><span class="si">}</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-138" name="__codelineno-0-138"></a>        <span class="k">return</span> <span class="n">vinterpretation</span>
<a id="__codelineno-0-139" name="__codelineno-0-139"></a>
<a id="__codelineno-0-140" name="__codelineno-0-140"></a>    <span class="c1"># provide static constructors for most important structural variation types</span>
<a id="__codelineno-0-141" name="__codelineno-0-141"></a>
<a id="__codelineno-0-142" name="__codelineno-0-142"></a>    <span class="nd">@staticmethod</span>
<a id="__codelineno-0-143" name="__codelineno-0-143"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">chromosomal_deletion</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-144" name="__codelineno-0-144"></a>                             <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-145" name="__codelineno-0-145"></a>                             <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146"></a>                             <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-147" name="__codelineno-0-147"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-148" name="__codelineno-0-148"></a><span class="sd">        create a StructuralVariant object for a chromosomal deletion</span>
<a id="__codelineno-0-149" name="__codelineno-0-149"></a>
<a id="__codelineno-0-150" name="__codelineno-0-150"></a><span class="sd">        :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-151" name="__codelineno-0-151"></a><span class="sd">        :type cell_contents: str</span>
<a id="__codelineno-0-152" name="__codelineno-0-152"></a><span class="sd">        :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-153" name="__codelineno-0-153"></a><span class="sd">        :type gene_symbol: str</span>
<a id="__codelineno-0-154" name="__codelineno-0-154"></a><span class="sd">        :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-155" name="__codelineno-0-155"></a><span class="sd">        :type gene_id: str</span>
<a id="__codelineno-0-156" name="__codelineno-0-156"></a><span class="sd">        :param variant_id: an identifier for the variant</span>
<a id="__codelineno-0-157" name="__codelineno-0-157"></a><span class="sd">        :type variant_id: str, optional</span>
<a id="__codelineno-0-158" name="__codelineno-0-158"></a><span class="sd">        &quot;&quot;&quot;</span>
<a id="__codelineno-0-159" name="__codelineno-0-159"></a>        <span class="k">return</span> <span class="n">StructuralVariant</span><span class="p">(</span><span class="n">cell_contents</span><span class="o">=</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-160" name="__codelineno-0-160"></a>                                 <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-161" name="__codelineno-0-161"></a>                                 <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-162" name="__codelineno-0-162"></a>                                 <span class="n">sequence_ontology_id</span><span class="o">=</span><span class="s2">&quot;SO:1000029&quot;</span><span class="p">,</span>
<a id="__codelineno-0-163" name="__codelineno-0-163"></a>                                 <span class="n">sequence_ontology_label</span><span class="o">=</span><span class="s2">&quot;chromosomal_deletion&quot;</span><span class="p">,</span>
<a id="__codelineno-0-164" name="__codelineno-0-164"></a>                                 <span class="n">variant_id</span><span class="o">=</span><span class="n">variant_id</span><span class="p">)</span>
<a id="__codelineno-0-165" name="__codelineno-0-165"></a>
<a id="__codelineno-0-166" name="__codelineno-0-166"></a>    <span class="nd">@staticmethod</span>
<a id="__codelineno-0-167" name="__codelineno-0-167"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">chromosomal_duplication</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-168" name="__codelineno-0-168"></a>                                <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-169" name="__codelineno-0-169"></a>                                <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-170" name="__codelineno-0-170"></a>                                <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-171" name="__codelineno-0-171"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-172" name="__codelineno-0-172"></a><span class="sd">        create a StructuralVariant object for a chromosomal duplication</span>
<a id="__codelineno-0-173" name="__codelineno-0-173"></a>
<a id="__codelineno-0-174" name="__codelineno-0-174"></a><span class="sd">        :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-175" name="__codelineno-0-175"></a><span class="sd">        :type cell_contents: str</span>
<a id="__codelineno-0-176" name="__codelineno-0-176"></a><span class="sd">        :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-177" name="__codelineno-0-177"></a><span class="sd">        :type gene_symbol: str</span>
<a id="__codelineno-0-178" name="__codelineno-0-178"></a><span class="sd">        :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-179" name="__codelineno-0-179"></a><span class="sd">        :type gene_id: str</span>
<a id="__codelineno-0-180" name="__codelineno-0-180"></a><span class="sd">        :param variant_id: an identifier for the variant</span>
<a id="__codelineno-0-181" name="__codelineno-0-181"></a><span class="sd">        :type variant_id: str, optional</span>
<a id="__codelineno-0-182" name="__codelineno-0-182"></a><span class="sd">        &quot;&quot;&quot;</span>
<a id="__codelineno-0-183" name="__codelineno-0-183"></a>        <span class="k">return</span> <span class="n">StructuralVariant</span><span class="p">(</span><span class="n">cell_contents</span><span class="o">=</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-184" name="__codelineno-0-184"></a>                                 <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-185" name="__codelineno-0-185"></a>                                 <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-186" name="__codelineno-0-186"></a>                                 <span class="n">sequence_ontology_id</span><span class="o">=</span><span class="s2">&quot;SO:1000037&quot;</span><span class="p">,</span>
<a id="__codelineno-0-187" name="__codelineno-0-187"></a>                                 <span class="n">sequence_ontology_label</span><span class="o">=</span><span class="s2">&quot;chromosomal_duplication&quot;</span><span class="p">,</span>
<a id="__codelineno-0-188" name="__codelineno-0-188"></a>                                 <span class="n">variant_id</span><span class="o">=</span><span class="n">variant_id</span><span class="p">)</span>
<a id="__codelineno-0-189" name="__codelineno-0-189"></a>
<a id="__codelineno-0-190" name="__codelineno-0-190"></a>    <span class="nd">@staticmethod</span>
<a id="__codelineno-0-191" name="__codelineno-0-191"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">chromosomal_inversion</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-192" name="__codelineno-0-192"></a>                              <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-193" name="__codelineno-0-193"></a>                              <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-194" name="__codelineno-0-194"></a>                              <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-195" name="__codelineno-0-195"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-196" name="__codelineno-0-196"></a><span class="sd">        create a StructuralVariant object for a chromosomal inversion</span>
<a id="__codelineno-0-197" name="__codelineno-0-197"></a>
<a id="__codelineno-0-198" name="__codelineno-0-198"></a><span class="sd">        :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-199" name="__codelineno-0-199"></a><span class="sd">        :type cell_contents: str</span>
<a id="__codelineno-0-200" name="__codelineno-0-200"></a><span class="sd">        :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-201" name="__codelineno-0-201"></a><span class="sd">        :type gene_symbol: str</span>
<a id="__codelineno-0-202" name="__codelineno-0-202"></a><span class="sd">        :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-203" name="__codelineno-0-203"></a><span class="sd">        :type gene_id: str</span>
<a id="__codelineno-0-204" name="__codelineno-0-204"></a><span class="sd">        :param variant_id: an identifier for the variant</span>
<a id="__codelineno-0-205" name="__codelineno-0-205"></a><span class="sd">        :type variant_id: str, optional</span>
<a id="__codelineno-0-206" name="__codelineno-0-206"></a><span class="sd">        &quot;&quot;&quot;</span>
<a id="__codelineno-0-207" name="__codelineno-0-207"></a>        <span class="k">return</span> <span class="n">StructuralVariant</span><span class="p">(</span><span class="n">cell_contents</span><span class="o">=</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-208" name="__codelineno-0-208"></a>                                 <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-209" name="__codelineno-0-209"></a>                                 <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-210" name="__codelineno-0-210"></a>                                 <span class="n">sequence_ontology_id</span><span class="o">=</span><span class="s2">&quot;SO:1000030&quot;</span><span class="p">,</span>
<a id="__codelineno-0-211" name="__codelineno-0-211"></a>                                 <span class="n">sequence_ontology_label</span><span class="o">=</span><span class="s2">&quot;chromosomal_inversion&quot;</span><span class="p">,</span>
<a id="__codelineno-0-212" name="__codelineno-0-212"></a>                                 <span class="n">variant_id</span><span class="o">=</span><span class="n">variant_id</span><span class="p">)</span>
<a id="__codelineno-0-213" name="__codelineno-0-213"></a>
<a id="__codelineno-0-214" name="__codelineno-0-214"></a>    <span class="nd">@staticmethod</span>
<a id="__codelineno-0-215" name="__codelineno-0-215"></a>    <span class="k">def</span><span class="w"> </span><span class="nf">chromosomal_translocation</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-216" name="__codelineno-0-216"></a>                              <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-217" name="__codelineno-0-217"></a>                              <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-218" name="__codelineno-0-218"></a>                              <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-219" name="__codelineno-0-219"></a><span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-220" name="__codelineno-0-220"></a><span class="sd">        create a StructuralVariant object for a chromosomal translocation</span>
<a id="__codelineno-0-221" name="__codelineno-0-221"></a>
<a id="__codelineno-0-222" name="__codelineno-0-222"></a><span class="sd">        :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-223" name="__codelineno-0-223"></a><span class="sd">        :type cell_contents: str</span>
<a id="__codelineno-0-224" name="__codelineno-0-224"></a><span class="sd">        :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-225" name="__codelineno-0-225"></a><span class="sd">        :type gene_symbol: str</span>
<a id="__codelineno-0-226" name="__codelineno-0-226"></a><span class="sd">        :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-227" name="__codelineno-0-227"></a><span class="sd">        :type gene_id: str</span>
<a id="__codelineno-0-228" name="__codelineno-0-228"></a><span class="sd">        :param variant_id: an identifier for the variant</span>
<a id="__codelineno-0-229" name="__codelineno-0-229"></a><span class="sd">        :type variant_id: str, optional</span>
<a id="__codelineno-0-230" name="__codelineno-0-230"></a><span class="sd">        &quot;&quot;&quot;</span>
<a id="__codelineno-0-231" name="__codelineno-0-231"></a>        <span class="k">return</span> <span class="n">StructuralVariant</span><span class="p">(</span><span class="n">cell_contents</span><span class="o">=</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-232" name="__codelineno-0-232"></a>                                 <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-233" name="__codelineno-0-233"></a>                                 <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-234" name="__codelineno-0-234"></a>                                 <span class="n">sequence_ontology_id</span><span class="o">=</span><span class="s2">&quot;SO:1000044&quot;</span><span class="p">,</span>
<a id="__codelineno-0-235" name="__codelineno-0-235"></a>                                 <span class="n">sequence_ontology_label</span><span class="o">=</span><span class="s2">&quot;chromosomal_translocation&quot;</span><span class="p">,</span>
<a id="__codelineno-0-236" name="__codelineno-0-236"></a>                                 <span class="n">variant_id</span><span class="o">=</span><span class="n">variant_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="pyphetools.creation.StructuralVariant.chromosomal_deletion" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">chromosomal_deletion</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h2>


    <div class="doc doc-contents ">

        <p>create a StructuralVariant object for a chromosomal deletion</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>cell_contents</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the string from the original table that we want to map as a structural variant</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_symbol</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the gene affected by the structural variant, e.g., GLI3</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_id</code>
            </td>
            <td>
                  <code>str, optional</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>an identifier for the variant</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>pyphetools/creation/structural_variant.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-142">142</a></span>
<span class="normal"><a href="#__codelineno-0-143">143</a></span>
<span class="normal"><a href="#__codelineno-0-144">144</a></span>
<span class="normal"><a href="#__codelineno-0-145">145</a></span>
<span class="normal"><a href="#__codelineno-0-146">146</a></span>
<span class="normal"><a href="#__codelineno-0-147">147</a></span>
<span class="normal"><a href="#__codelineno-0-148">148</a></span>
<span class="normal"><a href="#__codelineno-0-149">149</a></span>
<span class="normal"><a href="#__codelineno-0-150">150</a></span>
<span class="normal"><a href="#__codelineno-0-151">151</a></span>
<span class="normal"><a href="#__codelineno-0-152">152</a></span>
<span class="normal"><a href="#__codelineno-0-153">153</a></span>
<span class="normal"><a href="#__codelineno-0-154">154</a></span>
<span class="normal"><a href="#__codelineno-0-155">155</a></span>
<span class="normal"><a href="#__codelineno-0-156">156</a></span>
<span class="normal"><a href="#__codelineno-0-157">157</a></span>
<span class="normal"><a href="#__codelineno-0-158">158</a></span>
<span class="normal"><a href="#__codelineno-0-159">159</a></span>
<span class="normal"><a href="#__codelineno-0-160">160</a></span>
<span class="normal"><a href="#__codelineno-0-161">161</a></span>
<span class="normal"><a href="#__codelineno-0-162">162</a></span>
<span class="normal"><a href="#__codelineno-0-163">163</a></span>
<span class="normal"><a href="#__codelineno-0-164">164</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-142" name="__codelineno-0-142"></a><span class="nd">@staticmethod</span>
<a id="__codelineno-0-143" name="__codelineno-0-143"></a><span class="k">def</span><span class="w"> </span><span class="nf">chromosomal_deletion</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-144" name="__codelineno-0-144"></a>                         <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-145" name="__codelineno-0-145"></a>                         <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-146" name="__codelineno-0-146"></a>                         <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-147" name="__codelineno-0-147"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-148" name="__codelineno-0-148"></a><span class="sd">    create a StructuralVariant object for a chromosomal deletion</span>
<a id="__codelineno-0-149" name="__codelineno-0-149"></a>
<a id="__codelineno-0-150" name="__codelineno-0-150"></a><span class="sd">    :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-151" name="__codelineno-0-151"></a><span class="sd">    :type cell_contents: str</span>
<a id="__codelineno-0-152" name="__codelineno-0-152"></a><span class="sd">    :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-153" name="__codelineno-0-153"></a><span class="sd">    :type gene_symbol: str</span>
<a id="__codelineno-0-154" name="__codelineno-0-154"></a><span class="sd">    :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-155" name="__codelineno-0-155"></a><span class="sd">    :type gene_id: str</span>
<a id="__codelineno-0-156" name="__codelineno-0-156"></a><span class="sd">    :param variant_id: an identifier for the variant</span>
<a id="__codelineno-0-157" name="__codelineno-0-157"></a><span class="sd">    :type variant_id: str, optional</span>
<a id="__codelineno-0-158" name="__codelineno-0-158"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-159" name="__codelineno-0-159"></a>    <span class="k">return</span> <span class="n">StructuralVariant</span><span class="p">(</span><span class="n">cell_contents</span><span class="o">=</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-160" name="__codelineno-0-160"></a>                             <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-161" name="__codelineno-0-161"></a>                             <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-162" name="__codelineno-0-162"></a>                             <span class="n">sequence_ontology_id</span><span class="o">=</span><span class="s2">&quot;SO:1000029&quot;</span><span class="p">,</span>
<a id="__codelineno-0-163" name="__codelineno-0-163"></a>                             <span class="n">sequence_ontology_label</span><span class="o">=</span><span class="s2">&quot;chromosomal_deletion&quot;</span><span class="p">,</span>
<a id="__codelineno-0-164" name="__codelineno-0-164"></a>                             <span class="n">variant_id</span><span class="o">=</span><span class="n">variant_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="pyphetools.creation.StructuralVariant.chromosomal_duplication" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">chromosomal_duplication</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h2>


    <div class="doc doc-contents ">

        <p>create a StructuralVariant object for a chromosomal duplication</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>cell_contents</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the string from the original table that we want to map as a structural variant</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_symbol</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the gene affected by the structural variant, e.g., GLI3</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_id</code>
            </td>
            <td>
                  <code>str, optional</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>an identifier for the variant</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>pyphetools/creation/structural_variant.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-166">166</a></span>
<span class="normal"><a href="#__codelineno-0-167">167</a></span>
<span class="normal"><a href="#__codelineno-0-168">168</a></span>
<span class="normal"><a href="#__codelineno-0-169">169</a></span>
<span class="normal"><a href="#__codelineno-0-170">170</a></span>
<span class="normal"><a href="#__codelineno-0-171">171</a></span>
<span class="normal"><a href="#__codelineno-0-172">172</a></span>
<span class="normal"><a href="#__codelineno-0-173">173</a></span>
<span class="normal"><a href="#__codelineno-0-174">174</a></span>
<span class="normal"><a href="#__codelineno-0-175">175</a></span>
<span class="normal"><a href="#__codelineno-0-176">176</a></span>
<span class="normal"><a href="#__codelineno-0-177">177</a></span>
<span class="normal"><a href="#__codelineno-0-178">178</a></span>
<span class="normal"><a href="#__codelineno-0-179">179</a></span>
<span class="normal"><a href="#__codelineno-0-180">180</a></span>
<span class="normal"><a href="#__codelineno-0-181">181</a></span>
<span class="normal"><a href="#__codelineno-0-182">182</a></span>
<span class="normal"><a href="#__codelineno-0-183">183</a></span>
<span class="normal"><a href="#__codelineno-0-184">184</a></span>
<span class="normal"><a href="#__codelineno-0-185">185</a></span>
<span class="normal"><a href="#__codelineno-0-186">186</a></span>
<span class="normal"><a href="#__codelineno-0-187">187</a></span>
<span class="normal"><a href="#__codelineno-0-188">188</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-166" name="__codelineno-0-166"></a><span class="nd">@staticmethod</span>
<a id="__codelineno-0-167" name="__codelineno-0-167"></a><span class="k">def</span><span class="w"> </span><span class="nf">chromosomal_duplication</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-168" name="__codelineno-0-168"></a>                            <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-169" name="__codelineno-0-169"></a>                            <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-170" name="__codelineno-0-170"></a>                            <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-171" name="__codelineno-0-171"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-172" name="__codelineno-0-172"></a><span class="sd">    create a StructuralVariant object for a chromosomal duplication</span>
<a id="__codelineno-0-173" name="__codelineno-0-173"></a>
<a id="__codelineno-0-174" name="__codelineno-0-174"></a><span class="sd">    :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-175" name="__codelineno-0-175"></a><span class="sd">    :type cell_contents: str</span>
<a id="__codelineno-0-176" name="__codelineno-0-176"></a><span class="sd">    :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-177" name="__codelineno-0-177"></a><span class="sd">    :type gene_symbol: str</span>
<a id="__codelineno-0-178" name="__codelineno-0-178"></a><span class="sd">    :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-179" name="__codelineno-0-179"></a><span class="sd">    :type gene_id: str</span>
<a id="__codelineno-0-180" name="__codelineno-0-180"></a><span class="sd">    :param variant_id: an identifier for the variant</span>
<a id="__codelineno-0-181" name="__codelineno-0-181"></a><span class="sd">    :type variant_id: str, optional</span>
<a id="__codelineno-0-182" name="__codelineno-0-182"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-183" name="__codelineno-0-183"></a>    <span class="k">return</span> <span class="n">StructuralVariant</span><span class="p">(</span><span class="n">cell_contents</span><span class="o">=</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-184" name="__codelineno-0-184"></a>                             <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-185" name="__codelineno-0-185"></a>                             <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-186" name="__codelineno-0-186"></a>                             <span class="n">sequence_ontology_id</span><span class="o">=</span><span class="s2">&quot;SO:1000037&quot;</span><span class="p">,</span>
<a id="__codelineno-0-187" name="__codelineno-0-187"></a>                             <span class="n">sequence_ontology_label</span><span class="o">=</span><span class="s2">&quot;chromosomal_duplication&quot;</span><span class="p">,</span>
<a id="__codelineno-0-188" name="__codelineno-0-188"></a>                             <span class="n">variant_id</span><span class="o">=</span><span class="n">variant_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="pyphetools.creation.StructuralVariant.chromosomal_inversion" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">chromosomal_inversion</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h2>


    <div class="doc doc-contents ">

        <p>create a StructuralVariant object for a chromosomal inversion</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>cell_contents</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the string from the original table that we want to map as a structural variant</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_symbol</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the gene affected by the structural variant, e.g., GLI3</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_id</code>
            </td>
            <td>
                  <code>str, optional</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>an identifier for the variant</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>pyphetools/creation/structural_variant.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-190">190</a></span>
<span class="normal"><a href="#__codelineno-0-191">191</a></span>
<span class="normal"><a href="#__codelineno-0-192">192</a></span>
<span class="normal"><a href="#__codelineno-0-193">193</a></span>
<span class="normal"><a href="#__codelineno-0-194">194</a></span>
<span class="normal"><a href="#__codelineno-0-195">195</a></span>
<span class="normal"><a href="#__codelineno-0-196">196</a></span>
<span class="normal"><a href="#__codelineno-0-197">197</a></span>
<span class="normal"><a href="#__codelineno-0-198">198</a></span>
<span class="normal"><a href="#__codelineno-0-199">199</a></span>
<span class="normal"><a href="#__codelineno-0-200">200</a></span>
<span class="normal"><a href="#__codelineno-0-201">201</a></span>
<span class="normal"><a href="#__codelineno-0-202">202</a></span>
<span class="normal"><a href="#__codelineno-0-203">203</a></span>
<span class="normal"><a href="#__codelineno-0-204">204</a></span>
<span class="normal"><a href="#__codelineno-0-205">205</a></span>
<span class="normal"><a href="#__codelineno-0-206">206</a></span>
<span class="normal"><a href="#__codelineno-0-207">207</a></span>
<span class="normal"><a href="#__codelineno-0-208">208</a></span>
<span class="normal"><a href="#__codelineno-0-209">209</a></span>
<span class="normal"><a href="#__codelineno-0-210">210</a></span>
<span class="normal"><a href="#__codelineno-0-211">211</a></span>
<span class="normal"><a href="#__codelineno-0-212">212</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-190" name="__codelineno-0-190"></a><span class="nd">@staticmethod</span>
<a id="__codelineno-0-191" name="__codelineno-0-191"></a><span class="k">def</span><span class="w"> </span><span class="nf">chromosomal_inversion</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-192" name="__codelineno-0-192"></a>                          <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-193" name="__codelineno-0-193"></a>                          <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-194" name="__codelineno-0-194"></a>                          <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-195" name="__codelineno-0-195"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-196" name="__codelineno-0-196"></a><span class="sd">    create a StructuralVariant object for a chromosomal inversion</span>
<a id="__codelineno-0-197" name="__codelineno-0-197"></a>
<a id="__codelineno-0-198" name="__codelineno-0-198"></a><span class="sd">    :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-199" name="__codelineno-0-199"></a><span class="sd">    :type cell_contents: str</span>
<a id="__codelineno-0-200" name="__codelineno-0-200"></a><span class="sd">    :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-201" name="__codelineno-0-201"></a><span class="sd">    :type gene_symbol: str</span>
<a id="__codelineno-0-202" name="__codelineno-0-202"></a><span class="sd">    :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-203" name="__codelineno-0-203"></a><span class="sd">    :type gene_id: str</span>
<a id="__codelineno-0-204" name="__codelineno-0-204"></a><span class="sd">    :param variant_id: an identifier for the variant</span>
<a id="__codelineno-0-205" name="__codelineno-0-205"></a><span class="sd">    :type variant_id: str, optional</span>
<a id="__codelineno-0-206" name="__codelineno-0-206"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-207" name="__codelineno-0-207"></a>    <span class="k">return</span> <span class="n">StructuralVariant</span><span class="p">(</span><span class="n">cell_contents</span><span class="o">=</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-208" name="__codelineno-0-208"></a>                             <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-209" name="__codelineno-0-209"></a>                             <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-210" name="__codelineno-0-210"></a>                             <span class="n">sequence_ontology_id</span><span class="o">=</span><span class="s2">&quot;SO:1000030&quot;</span><span class="p">,</span>
<a id="__codelineno-0-211" name="__codelineno-0-211"></a>                             <span class="n">sequence_ontology_label</span><span class="o">=</span><span class="s2">&quot;chromosomal_inversion&quot;</span><span class="p">,</span>
<a id="__codelineno-0-212" name="__codelineno-0-212"></a>                             <span class="n">variant_id</span><span class="o">=</span><span class="n">variant_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="pyphetools.creation.StructuralVariant.chromosomal_translocation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">chromosomal_translocation</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h2>


    <div class="doc doc-contents ">

        <p>create a StructuralVariant object for a chromosomal translocation</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>cell_contents</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the string from the original table that we want to map as a structural variant</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_symbol</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the gene affected by the structural variant, e.g., GLI3</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_id</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_id</code>
            </td>
            <td>
                  <code>str, optional</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>an identifier for the variant</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>pyphetools/creation/structural_variant.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-214">214</a></span>
<span class="normal"><a href="#__codelineno-0-215">215</a></span>
<span class="normal"><a href="#__codelineno-0-216">216</a></span>
<span class="normal"><a href="#__codelineno-0-217">217</a></span>
<span class="normal"><a href="#__codelineno-0-218">218</a></span>
<span class="normal"><a href="#__codelineno-0-219">219</a></span>
<span class="normal"><a href="#__codelineno-0-220">220</a></span>
<span class="normal"><a href="#__codelineno-0-221">221</a></span>
<span class="normal"><a href="#__codelineno-0-222">222</a></span>
<span class="normal"><a href="#__codelineno-0-223">223</a></span>
<span class="normal"><a href="#__codelineno-0-224">224</a></span>
<span class="normal"><a href="#__codelineno-0-225">225</a></span>
<span class="normal"><a href="#__codelineno-0-226">226</a></span>
<span class="normal"><a href="#__codelineno-0-227">227</a></span>
<span class="normal"><a href="#__codelineno-0-228">228</a></span>
<span class="normal"><a href="#__codelineno-0-229">229</a></span>
<span class="normal"><a href="#__codelineno-0-230">230</a></span>
<span class="normal"><a href="#__codelineno-0-231">231</a></span>
<span class="normal"><a href="#__codelineno-0-232">232</a></span>
<span class="normal"><a href="#__codelineno-0-233">233</a></span>
<span class="normal"><a href="#__codelineno-0-234">234</a></span>
<span class="normal"><a href="#__codelineno-0-235">235</a></span>
<span class="normal"><a href="#__codelineno-0-236">236</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-214" name="__codelineno-0-214"></a><span class="nd">@staticmethod</span>
<a id="__codelineno-0-215" name="__codelineno-0-215"></a><span class="k">def</span><span class="w"> </span><span class="nf">chromosomal_translocation</span><span class="p">(</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-216" name="__codelineno-0-216"></a>                          <span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-217" name="__codelineno-0-217"></a>                          <span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-218" name="__codelineno-0-218"></a>                          <span class="n">variant_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-219" name="__codelineno-0-219"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-220" name="__codelineno-0-220"></a><span class="sd">    create a StructuralVariant object for a chromosomal translocation</span>
<a id="__codelineno-0-221" name="__codelineno-0-221"></a>
<a id="__codelineno-0-222" name="__codelineno-0-222"></a><span class="sd">    :param cell_contents: the string from the original table that we want to map as a structural variant</span>
<a id="__codelineno-0-223" name="__codelineno-0-223"></a><span class="sd">    :type cell_contents: str</span>
<a id="__codelineno-0-224" name="__codelineno-0-224"></a><span class="sd">    :param gene_symbol: the gene affected by the structural variant, e.g., GLI3</span>
<a id="__codelineno-0-225" name="__codelineno-0-225"></a><span class="sd">    :type gene_symbol: str</span>
<a id="__codelineno-0-226" name="__codelineno-0-226"></a><span class="sd">    :param gene_id: the identifier (using HGNC) of the gene, e.g., GLI3 is HGNC:4319</span>
<a id="__codelineno-0-227" name="__codelineno-0-227"></a><span class="sd">    :type gene_id: str</span>
<a id="__codelineno-0-228" name="__codelineno-0-228"></a><span class="sd">    :param variant_id: an identifier for the variant</span>
<a id="__codelineno-0-229" name="__codelineno-0-229"></a><span class="sd">    :type variant_id: str, optional</span>
<a id="__codelineno-0-230" name="__codelineno-0-230"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-231" name="__codelineno-0-231"></a>    <span class="k">return</span> <span class="n">StructuralVariant</span><span class="p">(</span><span class="n">cell_contents</span><span class="o">=</span><span class="n">cell_contents</span><span class="p">,</span>
<a id="__codelineno-0-232" name="__codelineno-0-232"></a>                             <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
<a id="__codelineno-0-233" name="__codelineno-0-233"></a>                             <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
<a id="__codelineno-0-234" name="__codelineno-0-234"></a>                             <span class="n">sequence_ontology_id</span><span class="o">=</span><span class="s2">&quot;SO:1000044&quot;</span><span class="p">,</span>
<a id="__codelineno-0-235" name="__codelineno-0-235"></a>                             <span class="n">sequence_ontology_label</span><span class="o">=</span><span class="s2">&quot;chromosomal_translocation&quot;</span><span class="p">,</span>
<a id="__codelineno-0-236" name="__codelineno-0-236"></a>                             <span class="n">variant_id</span><span class="o">=</span><span class="n">variant_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="pyphetools.creation.StructuralVariant.to_ga4gh_variant_interpretation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_ga4gh_variant_interpretation</span><span class="p">(</span><span class="n">acmg</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Transform this Variant object into a "variantInterpretation" message of the GA4GH Phenopacket schema</p>

            <details class="quote">
              <summary>Source code in <code>pyphetools/creation/structural_variant.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-69"> 69</a></span>
<span class="normal"><a href="#__codelineno-0-70"> 70</a></span>
<span class="normal"><a href="#__codelineno-0-71"> 71</a></span>
<span class="normal"><a href="#__codelineno-0-72"> 72</a></span>
<span class="normal"><a href="#__codelineno-0-73"> 73</a></span>
<span class="normal"><a href="#__codelineno-0-74"> 74</a></span>
<span class="normal"><a href="#__codelineno-0-75"> 75</a></span>
<span class="normal"><a href="#__codelineno-0-76"> 76</a></span>
<span class="normal"><a href="#__codelineno-0-77"> 77</a></span>
<span class="normal"><a href="#__codelineno-0-78"> 78</a></span>
<span class="normal"><a href="#__codelineno-0-79"> 79</a></span>
<span class="normal"><a href="#__codelineno-0-80"> 80</a></span>
<span class="normal"><a href="#__codelineno-0-81"> 81</a></span>
<span class="normal"><a href="#__codelineno-0-82"> 82</a></span>
<span class="normal"><a href="#__codelineno-0-83"> 83</a></span>
<span class="normal"><a href="#__codelineno-0-84"> 84</a></span>
<span class="normal"><a href="#__codelineno-0-85"> 85</a></span>
<span class="normal"><a href="#__codelineno-0-86"> 86</a></span>
<span class="normal"><a href="#__codelineno-0-87"> 87</a></span>
<span class="normal"><a href="#__codelineno-0-88"> 88</a></span>
<span class="normal"><a href="#__codelineno-0-89"> 89</a></span>
<span class="normal"><a href="#__codelineno-0-90"> 90</a></span>
<span class="normal"><a href="#__codelineno-0-91"> 91</a></span>
<span class="normal"><a href="#__codelineno-0-92"> 92</a></span>
<span class="normal"><a href="#__codelineno-0-93"> 93</a></span>
<span class="normal"><a href="#__codelineno-0-94"> 94</a></span>
<span class="normal"><a href="#__codelineno-0-95"> 95</a></span>
<span class="normal"><a href="#__codelineno-0-96"> 96</a></span>
<span class="normal"><a href="#__codelineno-0-97"> 97</a></span>
<span class="normal"><a href="#__codelineno-0-98"> 98</a></span>
<span class="normal"><a href="#__codelineno-0-99"> 99</a></span>
<span class="normal"><a href="#__codelineno-0-100">100</a></span>
<span class="normal"><a href="#__codelineno-0-101">101</a></span>
<span class="normal"><a href="#__codelineno-0-102">102</a></span>
<span class="normal"><a href="#__codelineno-0-103">103</a></span>
<span class="normal"><a href="#__codelineno-0-104">104</a></span>
<span class="normal"><a href="#__codelineno-0-105">105</a></span>
<span class="normal"><a href="#__codelineno-0-106">106</a></span>
<span class="normal"><a href="#__codelineno-0-107">107</a></span>
<span class="normal"><a href="#__codelineno-0-108">108</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-69" name="__codelineno-0-69"></a><span class="k">def</span><span class="w"> </span><span class="nf">to_ga4gh_variant_interpretation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">acmg</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-70" name="__codelineno-0-70"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-71" name="__codelineno-0-71"></a><span class="sd">    Transform this Variant object into a &quot;variantInterpretation&quot; message of the GA4GH Phenopacket schema</span>
<a id="__codelineno-0-72" name="__codelineno-0-72"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-73" name="__codelineno-0-73"></a>    <span class="n">vdescriptor</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">VariationDescriptor</span><span class="p">()</span>
<a id="__codelineno-0-74" name="__codelineno-0-74"></a>    <span class="n">vdescriptor</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_variant_id</span>
<a id="__codelineno-0-75" name="__codelineno-0-75"></a>    <span class="n">vdescriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">value_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hgnc_id</span>
<a id="__codelineno-0-76" name="__codelineno-0-76"></a>    <span class="n">vdescriptor</span><span class="o">.</span><span class="n">gene_context</span><span class="o">.</span><span class="n">symbol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_gene_symbol</span>
<a id="__codelineno-0-77" name="__codelineno-0-77"></a>    <span class="n">vdescriptor</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_label</span>
<a id="__codelineno-0-78" name="__codelineno-0-78"></a>    <span class="n">vdescriptor</span><span class="o">.</span><span class="n">molecule_context</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">MoleculeContext</span><span class="o">.</span><span class="n">genomic</span>
<a id="__codelineno-0-79" name="__codelineno-0-79"></a>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-80" name="__codelineno-0-80"></a>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="o">==</span> <span class="s1">&#39;heterozygous&#39;</span><span class="p">:</span>
<a id="__codelineno-0-81" name="__codelineno-0-81"></a>            <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;GENO:0000135&quot;</span>
<a id="__codelineno-0-82" name="__codelineno-0-82"></a>            <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;heterozygous&quot;</span>
<a id="__codelineno-0-83" name="__codelineno-0-83"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="o">==</span> <span class="s1">&#39;homozygous&#39;</span><span class="p">:</span>
<a id="__codelineno-0-84" name="__codelineno-0-84"></a>            <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;GENO:0000136&quot;</span>
<a id="__codelineno-0-85" name="__codelineno-0-85"></a>            <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;homozygous&quot;</span>
<a id="__codelineno-0-86" name="__codelineno-0-86"></a>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span> <span class="o">==</span> <span class="s1">&#39;hemizygous&#39;</span><span class="p">:</span>
<a id="__codelineno-0-87" name="__codelineno-0-87"></a>            <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s2">&quot;GENO:0000134&quot;</span>
<a id="__codelineno-0-88" name="__codelineno-0-88"></a>            <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;hemizygous&quot;</span>
<a id="__codelineno-0-89" name="__codelineno-0-89"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-90" name="__codelineno-0-90"></a>            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Did not recognize genotype </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-91" name="__codelineno-0-91"></a>    <span class="n">vdescriptor</span><span class="o">.</span><span class="n">structural_type</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_so_id</span>
<a id="__codelineno-0-92" name="__codelineno-0-92"></a>    <span class="n">vdescriptor</span><span class="o">.</span><span class="n">structural_type</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_so_label</span>
<a id="__codelineno-0-93" name="__codelineno-0-93"></a>    <span class="n">vinterpretation</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">VariantInterpretation</span><span class="p">()</span>
<a id="__codelineno-0-94" name="__codelineno-0-94"></a>    <span class="k">if</span> <span class="n">acmg</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-95" name="__codelineno-0-95"></a>        <span class="k">if</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;benign&#39;</span><span class="p">:</span>
<a id="__codelineno-0-96" name="__codelineno-0-96"></a>            <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">BENIGN</span>
<a id="__codelineno-0-97" name="__codelineno-0-97"></a>        <span class="k">elif</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span> <span class="o">==</span> <span class="s1">&#39;likely benign&#39;</span> <span class="ow">or</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;likely_benign&#39;</span><span class="p">:</span>
<a id="__codelineno-0-98" name="__codelineno-0-98"></a>            <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">LIKELY_BENIGN</span>
<a id="__codelineno-0-99" name="__codelineno-0-99"></a>        <span class="k">elif</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span> <span class="o">==</span> <span class="s1">&#39;uncertain significance&#39;</span> <span class="ow">or</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;uncertain_significance&#39;</span><span class="p">:</span>
<a id="__codelineno-0-100" name="__codelineno-0-100"></a>            <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">UNCERTAIN_SIGNIFICANCE</span>
<a id="__codelineno-0-101" name="__codelineno-0-101"></a>        <span class="k">elif</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span> <span class="o">==</span> <span class="s1">&#39;likely pathogenic&#39;</span> <span class="ow">or</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;likely_pathogenic&#39;</span><span class="p">:</span>
<a id="__codelineno-0-102" name="__codelineno-0-102"></a>            <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">LIKELY_PATHOGENIC</span>
<a id="__codelineno-0-103" name="__codelineno-0-103"></a>        <span class="k">elif</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span> <span class="o">==</span> <span class="s1">&#39;pathogenic&#39;</span> <span class="ow">or</span> <span class="n">acmg</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;pathogenic&#39;</span><span class="p">:</span>
<a id="__codelineno-0-104" name="__codelineno-0-104"></a>            <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">phenopackets</span><span class="o">.</span><span class="n">AcmgPathogenicityClassification</span><span class="o">.</span><span class="n">PATHOGENIC</span>
<a id="__codelineno-0-105" name="__codelineno-0-105"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-106" name="__codelineno-0-106"></a>            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning- did not recognize ACMG category </span><span class="si">{</span><span class="n">acmg</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-107" name="__codelineno-0-107"></a>    <span class="n">vinterpretation</span><span class="o">.</span><span class="n">variation_descriptor</span><span class="o">.</span><span class="n">CopyFrom</span><span class="p">(</span><span class="n">vdescriptor</span><span class="p">)</span>
<a id="__codelineno-0-108" name="__codelineno-0-108"></a>    <span class="k">return</span> <span class="n">vinterpretation</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="pyphetools.creation.StructuralVariant.to_variant_interpretation_202" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">to_variant_interpretation_202</span><span class="p">(</span><span class="n">acmg</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Transform this Variant object into a "variantInterpretation" message of the GA4GH Phenopacket schema</p>

            <details class="quote">
              <summary>Source code in <code>pyphetools/creation/structural_variant.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-110">110</a></span>
<span class="normal"><a href="#__codelineno-0-111">111</a></span>
<span class="normal"><a href="#__codelineno-0-112">112</a></span>
<span class="normal"><a href="#__codelineno-0-113">113</a></span>
<span class="normal"><a href="#__codelineno-0-114">114</a></span>
<span class="normal"><a href="#__codelineno-0-115">115</a></span>
<span class="normal"><a href="#__codelineno-0-116">116</a></span>
<span class="normal"><a href="#__codelineno-0-117">117</a></span>
<span class="normal"><a href="#__codelineno-0-118">118</a></span>
<span class="normal"><a href="#__codelineno-0-119">119</a></span>
<span class="normal"><a href="#__codelineno-0-120">120</a></span>
<span class="normal"><a href="#__codelineno-0-121">121</a></span>
<span class="normal"><a href="#__codelineno-0-122">122</a></span>
<span class="normal"><a href="#__codelineno-0-123">123</a></span>
<span class="normal"><a href="#__codelineno-0-124">124</a></span>
<span class="normal"><a href="#__codelineno-0-125">125</a></span>
<span class="normal"><a href="#__codelineno-0-126">126</a></span>
<span class="normal"><a href="#__codelineno-0-127">127</a></span>
<span class="normal"><a href="#__codelineno-0-128">128</a></span>
<span class="normal"><a href="#__codelineno-0-129">129</a></span>
<span class="normal"><a href="#__codelineno-0-130">130</a></span>
<span class="normal"><a href="#__codelineno-0-131">131</a></span>
<span class="normal"><a href="#__codelineno-0-132">132</a></span>
<span class="normal"><a href="#__codelineno-0-133">133</a></span>
<span class="normal"><a href="#__codelineno-0-134">134</a></span>
<span class="normal"><a href="#__codelineno-0-135">135</a></span>
<span class="normal"><a href="#__codelineno-0-136">136</a></span>
<span class="normal"><a href="#__codelineno-0-137">137</a></span>
<span class="normal"><a href="#__codelineno-0-138">138</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-110" name="__codelineno-0-110"></a><span class="k">def</span><span class="w"> </span><span class="nf">to_variant_interpretation_202</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">acmg</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-111" name="__codelineno-0-111"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-112" name="__codelineno-0-112"></a><span class="sd">    Transform this Variant object into a &quot;variantInterpretation&quot; message of the GA4GH Phenopacket schema</span>
<a id="__codelineno-0-113" name="__codelineno-0-113"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-114" name="__codelineno-0-114"></a>    <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">VariantInterpretation</span> <span class="k">as</span> <span class="n">VariantInterpretation202</span>
<a id="__codelineno-0-115" name="__codelineno-0-115"></a>    <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">VariationDescriptor</span> <span class="k">as</span> <span class="n">VariationDescriptor202</span>
<a id="__codelineno-0-116" name="__codelineno-0-116"></a>    <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">GeneDescriptor</span> <span class="k">as</span> <span class="n">GeneDescriptor202</span>
<a id="__codelineno-0-117" name="__codelineno-0-117"></a>    <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">OntologyClass</span> <span class="k">as</span> <span class="n">OntologyClass202</span>
<a id="__codelineno-0-118" name="__codelineno-0-118"></a>    <span class="kn">from</span><span class="w"> </span><span class="nn">..pp.v202</span><span class="w"> </span><span class="kn">import</span> <span class="n">MoleculeContext</span> <span class="k">as</span> <span class="n">MoleculeContext202</span>
<a id="__codelineno-0-119" name="__codelineno-0-119"></a>
<a id="__codelineno-0-120" name="__codelineno-0-120"></a>    <span class="n">gdesc</span> <span class="o">=</span> <span class="n">GeneDescriptor202</span><span class="p">(</span><span class="n">value_id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_hgnc_id</span><span class="p">,</span> <span class="n">symbol</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_gene_symbol</span><span class="p">)</span>
<a id="__codelineno-0-121" name="__codelineno-0-121"></a>    <span class="n">vdescriptor</span> <span class="o">=</span> <span class="n">VariationDescriptor202</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_variant_id</span><span class="p">,</span>
<a id="__codelineno-0-122" name="__codelineno-0-122"></a>                                         <span class="n">gene_context</span><span class="o">=</span><span class="n">gdesc</span><span class="p">,</span>
<a id="__codelineno-0-123" name="__codelineno-0-123"></a>                                         <span class="n">label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_label</span><span class="p">,</span>
<a id="__codelineno-0-124" name="__codelineno-0-124"></a>                                         <span class="n">molecule_context</span><span class="o">=</span><span class="n">MoleculeContext202</span><span class="o">.</span><span class="n">genomic</span><span class="p">)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125"></a>    <span class="c1"># Note that it is possible to first create a StructuralVariant object and later set its genotype</span>
<a id="__codelineno-0-126" name="__codelineno-0-126"></a>    <span class="c1"># Therefore, it is not an error if gt_term is None</span>
<a id="__codelineno-0-127" name="__codelineno-0-127"></a>    <span class="n">gt_term</span> <span class="o">=</span> <span class="n">Variant</span><span class="o">.</span><span class="n">_get_genotype_term</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genotype</span><span class="p">)</span>
<a id="__codelineno-0-128" name="__codelineno-0-128"></a>    <span class="k">if</span> <span class="n">gt_term</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-129" name="__codelineno-0-129"></a>        <span class="n">vdescriptor</span><span class="o">.</span><span class="n">allelic_state</span> <span class="o">=</span> <span class="n">gt_term</span>
<a id="__codelineno-0-130" name="__codelineno-0-130"></a>    <span class="n">structural_type</span> <span class="o">=</span> <span class="n">OntologyClass202</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_so_id</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_so_label</span><span class="p">)</span>
<a id="__codelineno-0-131" name="__codelineno-0-131"></a>    <span class="n">vdescriptor</span><span class="o">.</span><span class="n">structural_type</span> <span class="o">=</span> <span class="n">structural_type</span>
<a id="__codelineno-0-132" name="__codelineno-0-132"></a>    <span class="n">vinterpretation</span> <span class="o">=</span> <span class="n">VariantInterpretation202</span><span class="p">(</span><span class="n">variation_descriptor</span><span class="o">=</span><span class="n">vdescriptor</span><span class="p">)</span>
<a id="__codelineno-0-133" name="__codelineno-0-133"></a>    <span class="n">acmg_code</span> <span class="o">=</span> <span class="n">Variant</span><span class="o">.</span><span class="n">_get_acmg_classification</span><span class="p">(</span><span class="n">acmg</span><span class="o">=</span><span class="n">acmg</span><span class="p">)</span>
<a id="__codelineno-0-134" name="__codelineno-0-134"></a>    <span class="k">if</span> <span class="n">acmg_code</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-135" name="__codelineno-0-135"></a>        <span class="n">vinterpretation</span><span class="o">.</span><span class="n">acmgPathogenicityClassification</span> <span class="o">=</span> <span class="n">acmg_code</span>
<a id="__codelineno-0-136" name="__codelineno-0-136"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-137" name="__codelineno-0-137"></a>        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning- did not recognize ACMG category </span><span class="se">\&quot;</span><span class="si">{</span><span class="n">acmg</span><span class="si">}</span><span class="se">\&quot;</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-138" name="__codelineno-0-138"></a>    <span class="k">return</span> <span class="n">vinterpretation</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>







  
  






                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; 2023-2024 <a href="https://github.com/pnrobinson"  target="_blank" rel="noopener">Peter N Robinson</a>

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/pnrobinson" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://www.linkedin.com/in/peter-n-robinson-b7833811/" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3M135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5m282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.sections", "toc.integrate", "navigation.top", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy", "content.code.select"], "search": "../../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.88dd0f4e.min.js"></script>
      
    
  </body>
</html>